crates/provider-wadm/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+                                            }
+                                        };
+
+                                        let component_id = Arc::clone(&component_id);
+                                        let app_name = Arc::clone(&app_name);
+                                        let wrpc = Arc::clone(&wrpc);
+                                        tokio::spawn(async move {
+                                            dispatch_status_update(
+                                                &wrpc,
+                                                component_id.as_str(),
+                                                &app_name,
+                                                status.into(),
+                                                permit,
+                                            )
+                                            .instrument(span)
+                                            .await;
+                                        });
+                                    } else {
+                                        warn!("Received status OK but no status provided");
+                                    }
+                                }
+                            },
+                            Err(e) => {
+                                warn!("Failed to deserialize message: {}", e);
+                            }
+                        };
+                    }
+                }
+            },
+            abort_registration,
+        ));
+
+        Ok(abort_handle)
+    }
+
+    /// Helper function to get the NATS client from the context
+    async fn get_client(&self, ctx: Option<Context>) -> anyhow::Result<Client> {
+        if let Some(ref source_id) = ctx
+            .as_ref()
+            .and_then(|Context { component, .. }| component.clone())
+        {
+            let actors = self.consumer_components.read().await;
+            let wadm_bundle = match actors.get(source_id) {
+                Some(wadm_bundle) => wadm_bundle,
+                None => {
+                    error!("actor not linked: {source_id}");
+                    bail!("actor not linked: {source_id}")
+                }
+            };
+            Ok(wadm_bundle.client.clone())
+        } else {
+            error!("no actor in request");
+            bail!("no actor in request")
+        }
+    }
+}
+
+#[instrument(level = "debug", skip_all, fields(component_id = %component_id, app_name = %app))]
+async fn dispatch_status_update(
+    wrpc: &WrpcClient,
+    component_id: &str,
+    app: &str,
+    status: Status,
+    _permit: OwnedSemaphorePermit,
+) {
+    let update = StatusUpdate {
+        app: app.to_string(),
+        status,
+    };
+    debug!(
+        app = app,
+        component_id = component_id,
+        "sending status to component",
+    );
+
+    let mut cx = async_nats::HeaderMap::new();
+    for (k, v) in TraceContextInjector::default_with_span().iter() {
```

Comment:
Nit: you can use the `opentelemetry-nats` crate available in the workspace dependencies to do this!